Tests/Realm.Tests/Database/PrimaryKeyTests.cs

```diff
@@ -192,6 +195,57 @@ private RealmObjectBase FindByPKGeneric(Type type, object primaryKeyValue, PKTyp
             return (RealmObjectBase)genericMethod.MakeGenericMethod(type).Invoke(_realm, new[] { primaryKeyValue });
         }
 
+        [Test]
+        public void RealmFind_WhenPKIsIntAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<IntPrimaryKeyWithValueObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsCharAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyCharObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsByteAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyByteObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsInt16AndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyInt16Object>();
+
+        [Test]
+        public void RealmFind_WhenPKIsInt32AndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyInt32Object>();
+
+        [Test]
+        public void RealmFind_WhenPKIsInt64AndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyInt64Object>();
+
+        [Test]
+        public void RealmFind_WhenPKIsStringAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyStringObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsGuidAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyGuidObject>();
+
+        [Test]
+        public void RealmFind_WhenPKIsObjectIdAndArgumentIsNot_Throws() => RealmFind_IncorrectPKArgument_Throws<PrimaryKeyObjectIdObject>();
+
+        // In Find<T> anything that is a primitive type (according to Realm's view) is converted to long?.
+        // FindCore<T> converts its input in RealmValue. Which in turn treats long as PritiveValue of type Int.
+        // Without this pass it'd be impossible to match any PK that's of primitive type, since a PK defined as e.g. short would always be
+        // compared against an int
```

Comment:
I understand what you're trying to say here, but I think we should reword it to make it easier to understand for people who don't have context. I'm thinking of something like:

```
// In RealmFind_IncorrectPKArgument_Throws we're using reflection to check if the primary
// key type and the property type match. Since Realm stores all integral properties as Int64, we
// want to treat the narrower types as "long" to match the value type in _primaryKeyValues.
```